### PR TITLE
Add artifact and builder graphQL schema

### PR DESCRIPTION
**Pull request by pxp928**
Add artifact (algorithm and digest) and builder (uri) for graphQL
NOTE: Needs Rebase once https://github.com/guacsec/guac/pull/398 and https://github.com/guacsec/guac/pull/399 merge!
```
fragment allArtifactTree on Artifact {
  	algorithm
    digest
}

fragment allBuilderTree on Builder {
  uri
}


query Q1 {
  artifacts(artifactSpec: {algorithm: "sha256", digest: "6bbb0da1891646e58eb3e6a63af3a6fc3c8eb5a0d44824cba581d2e14a0450cf"}) {
    ...allArtifactTree
  }
}

query Q2 {
  artifacts(artifactSpec: {}) {
    ...allArtifactTree
  }
}

query Q3 {
  artifacts(artifactSpec: {digest: "6bbb0da1891646e58eb3e6a63af3a6fc3c8eb5a0d44824cba581d2e14a0450cf"}) {
    ...allArtifactTree
  }
}

query Q4 {
  artifacts(artifactSpec: {algorithm: "sha512"}) {
    ...allArtifactTree
  }
}

query Q5 {
  builders(builderSpec: {uri: "https://github.com/Attestations/GitHubHostedActions@v1"}) {
    ...allBuilderTree
  }
}

query Q6 {
  builders(builderSpec: {}) {
    ...allBuilderTree
  }
}
```

### CODE REVIEW

File: pkg/assembler/backends/testing/backend.go [hunk 1]
```diff
@@ -45,11 +68,97 @@ func (c *demoClient) Packages(ctx context.Context, pkgSpec *model.PkgSpec) ([]*m
...
+	for _, a := range c.artifacts {
+		if artifactSpec.Digest == nil && artifactSpec.Algorithm == nil {
+			artifacts = append(artifacts, a)
+		} else if artifactSpec.Digest != nil && artifactSpec.Algorithm == nil && a.Digest == *artifactSpec.Digest {
+			artifacts = append(artifacts, a)
+		} else if artifactSpec.Digest == nil && artifactSpec.Algorithm != nil && a.Algorithm == *artifactSpec.Algorithm {
+			artifacts = append(artifacts, a)
+		} else if artifactSpec.Digest != nil && artifactSpec.Algorithm != nil && a.Algorithm == *artifactSpec.Algorithm && a.Digest == *artifactSpec.Digest {
+			artifacts = append(artifacts, a)
+		}
```
Comment (lumjjb) [inline pkg/assembler/backends/testing/backend.go:142]: i think we need to decide on the behavior here on cap sensitivity both for the protocol or the digest. We could follow what the intoto recommendation is and do lower case everything and have that be checked on ingestion time, we would want to specify it here and probably have the testing backend bootstrapped artifacts follow that convention.

Comment (pxp928) [reply pkg/assembler/backends/testing/backend.go:142]: Yup, I agree to make it all lowercase. I will enforce lowercase in the testing.

Comment (mihaimaruseac) [reply pkg/assembler/backends/testing/backend.go:142]: +1, we need to lowercase both when receiving the query and on ingesting.

**Review by mihaimaruseac — APPROVED** [review]

File: pkg/assembler/backends/testing/backend.go (hunk 1 above)
Comment (pxp928) [reply pkg/assembler/backends/testing/backend.go]: addressed PTAL!

**Comment by pxp928** [comment]
Rebased! PTAL!

**Review by mihaimaruseac — APPROVED** [review]

File: pkg/assembler/backends/testing/ingest_artifact.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"strings"
+
+	"github.com/guacsec/guac/pkg/assembler/graphql/model"
+)
+
+func registerAllArtifacts(client *demoClient) {
+	// strings.ToLower(string(checksum.Algorithm)) + ":" + checksum.Value
+	client.registerArtifact("sha256", "6bbb0da1891646e58eb3e6a63af3a6fc3c8eb5a0d44824cba581d2e14a0450cf")
+	client.registerArtifact("sha1", "7A8F47318E4676DACB0142AFA0B83029CD7BEFD9")
```
Comment (lumjjb) [inline pkg/assembler/backends/testing/ingest_artifact.go:27]: still needs to be lowercase?

**Review by lumjjb — APPROVED** [review]
lgtm otherwise (besides testing backend case comments)

File: pkg/assembler/backends/testing/ingest_artifact.go (hunk 2 above)
Comment (pxp928) [reply pkg/assembler/backends/testing/ingest_artifact.go:27]: they are lowercased in the `registerArtifact` below

Comment (lumjjb) [reply pkg/assembler/backends/testing/ingest_artifact.go:27]: ah ok i was thinking that we should try to communicate that we want the input to be lowercase, but if we're going to have ingestion lower case it then we should be OK

**Pull request merged**